test/metabase/driver_test.clj

```diff
@@ -68,6 +75,48 @@
                        :tables
                        (some :schema))))))))
 
+(deftest check-can-connect-before-sync-test
+  (testing "Database sync should short-circuit and fail if the database connection is not available"
+    (mt/test-drivers (mt/normal-drivers)
+      (let [;; create a database with a table and sync
+            database-name      (name (gensym))
+            empty-dbdef        {:database-name database-name}
+            _                  (tx/create-db! driver/*driver* empty-dbdef)
+            connection-details (tx/dbdef->connection-details driver/*driver* :db empty-dbdef)
+            db                 (first (t2/insert-returning-instances! :model/Database {:name       database-name
+                                                                                       :engine     (u/qualified-name driver/*driver*)
+                                                                                       :details    connection-details}))
+            log-match?         (fn [[log-level throwable message]]
+                                 (and (= log-level :warn)
+                                      (instance? clojure.lang.ExceptionInfo throwable)
+                                      (re-matches #"^Cannot sync Database (.+): (.+)" message)))]
+        (binding [h2/*allow-testing-h2-connections* true]
+          (sync/sync-database! db))
+        (testing "sense checks before deleting the database"
+          (testing "sense check 1: sync-and-analyze-database! should not log a warning"
+            (is (not-any?
+                 log-match?
+                 (mt/with-log-messages-for-level :warn
+                   (#'task.sync-databases/sync-and-analyze-database!* (u/the-id db))))))
+          (testing "sense check 2: sync-schema should return true and start a sync"
+            (is (= {:status "ok"}
+                   (mt/user-http-request :crowberto :post 200 (str "/database/" (u/the-id db) "/sync_schema"))))))
+        ;; invalidate the connection pool so we don't have to wait to finish syncing before destroying the db
+        (when (isa? driver/hierarchy driver/*driver* :sql-jdbc)
+          (sql-jdbc.conn/invalidate-pool-for-db! db))
+        ;; delete the db
+        (tx/destroy-db! driver/*driver* empty-dbdef)
+        (testing "after deleting a database, sync should fail"
+          (testing "1: sync-and-analyze-database! should log a warning and fail early"
+            (is (some
+                 log-match?
+                 (mt/with-log-messages-for-level :warn
+                   (#'task.sync-databases/sync-and-analyze-database!* (u/the-id db))))))
+          (testing "2: triggering the sync via the POST /api/database/:id/sync_schema endpoint should fail"
+            (mt/user-http-request :crowberto :post 422 (str "/database/" (u/the-id db) "/sync_schema"))))
+        ;; clean up the database
+        (t2/delete! :model/Database (u/the-id db))))))
```

Comment:
use `mt/with-model-cleanup` instead?

Reply:
`mt/with-model-cleanup` doesn't execute the toucan hooks, which is needed here to clean up connection resources

https://github.com/metabase/metabase/blob/cb2ba50f3d4a7b7fc9f1634b27681514b561d940/test/metabase/test/util.clj#L663-L664